src/libraries/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -2058,69 +2093,112 @@ internal Node(TKey key, TValue value, int hashcode, Node next)
         {
-            private readonly IEnumerator<KeyValuePair<TKey, TValue>> _enumerator; // Enumerator over the dictionary.
+            /// <summary>A singly-linked list for each bucket.</summary>
+            internal readonly Node?[] _buckets;
+            /// <summary>A set of locks, each guarding a section of the table.</summary>
+            internal readonly object[] _locks;
+            /// <summary>The number of elements guarded by each lock.</summary>
+            internal readonly int[] _countPerLock;
+            /// <summary>Pre-computed multiplier for use on 64-bit performing faster modulo operations.</summary>
+            internal readonly ulong _fastModBucketsMultiplier;
 
-            internal DictionaryEnumerator(ConcurrentDictionary<TKey, TValue> dictionary)
+            internal Tables(Node?[] buckets, object[] locks, int[] countPerLock)
             {
-                _enumerator = dictionary.GetEnumerator();
+                _buckets = buckets;
+                _locks = locks;
+                _countPerLock = countPerLock;
+                if (IntPtr.Size == 8)
+                {
+                    _fastModBucketsMultiplier = HashHelpers.GetFastModMultiplier((uint)buckets.Length);
+                }
             }
 
-            public DictionaryEntry Entry
+            /// <summary>Computes a ref to the bucket for a particular key.</summary>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal ref Node? GetBucket(uint hashcode)
             {
-                get { return new DictionaryEntry(_enumerator.Current.Key, _enumerator.Current.Value); }
+                Node?[] buckets = _buckets;
+                if (IntPtr.Size == 8)
+                {
+                    return ref buckets[HashHelpers.FastMod(hashcode, (uint)buckets.Length, _fastModBucketsMultiplier)];
+                }
+                else
+                {
+                    return ref buckets[hashcode % (uint)buckets.Length];
+                }
             }
 
-            public object Key
+            /// <summary>Computes the bucket and lock number for a particular key.</summary>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal ref Node? GetBucketAndLock(uint hashcode, out uint lockNo)
             {
-                get { return _enumerator.Current.Key; }
+                Node?[] buckets = _buckets;
+                if (IntPtr.Size == 8)
+                {
+                    uint bucketNo = HashHelpers.FastMod(hashcode, (uint)buckets.Length, _fastModBucketsMultiplier);
+                    lockNo = bucketNo % (uint)_locks.Length;
+                    return ref buckets[bucketNo];
+                }
+                else
+                {
+                    uint bucketNo = hashcode % (uint)_buckets.Length;
+                    lockNo = bucketNo % (uint)_locks.Length;
+                    return ref buckets[bucketNo];
+                }
```

Comment:
It is interesting that you optimize only one of the two modulo operations. If that is intentional, could the two common lines be moved out of this `if`?

Reply:
This sizes are different so it would require two different multipliers to be stored, and this second one is only used on locking code paths where the difference is mostly noise, so it didn't seem worthwhile. I could look at moving out the last two lines, separating the declaration and the initialization of the bucketNo.

Reply:
Don't think it's worth it; but you could use only half of the `long` since it's already there (e.g. cast to `(uint)`). However I assume it will do a more expensive long read, then truncate it on 32bit rather than just doing a 32bit read...